### PR TITLE
  I have blacked all files and will now start a code review

### PR DESCRIPTION
**Pull request by mahrossi**
I had a look at the output of flake8 and there is still a lot to be addressed. @maksimovdmitrii can take care of that

### CODE REVIEW

File: gensec.py [hunk 1]
```diff
@@ -55,19 +62,29 @@
...
+    )
     workflow.success = dirs.dir_num
-    generated_dirs = [z for z in os.listdir(dirs.generate_folder) if os.path.isdir(os.path.join(dirs.generate_folder, z))]
+    generated_dirs = [
+        z
+        for z in os.listdir(dirs.generate_folder)
+        if os.path.isdir(os.path.join(dirs.generate_folder, z))
+    ]
 
     # while len(generated_dirs)>0:
```
Comment (mahrossi) [inline gensec.py:89]: Large portions of commented code should be cleaned out. if one really wants to save it for future purposes, do it in a branch.

File: gensec.py [hunk 2]
```diff
@@ -24,29 +24,36 @@
 from ase.io import write
```
Comment (mahrossi) [inline gensec.py:25]: This file is a very large piece of code with many if/else and no docstrings or documentation. This should at least include a much much more detailed docstring that documents the workflow that is being implemented.

File: gensec/customPrecon.py [hunk 3]
```diff
@@ -17,25 +17,26 @@
...
+# import time
 
 import numpy as np
 
 from ase.constraints import Filter, FixAtoms
 from ase.geometry.cell import cell_to_cellpar
 from ase.neighborlist import neighbor_list
 
+
 def Kabsh_rmsd(atoms, initial, molindixes, removeHs=False):
```
Comment (mahrossi) [inline gensec/customPrecon.py:29]: Docstring missing

File: gensec/customPrecon.py [hunk 4]
```diff
@@ -62,100 +63,335 @@ def Kabsh_rmsd(atoms, initial, molindixes, removeHs=False):
 #    !single double theory with accrate basis. The vdW radii for respective element are
 #    !defined as discussed in Tkatchenko, A. & Scheffler, M. Phys. Rev. Lett. 102, 073005 (2009).
 
-BOHR_to_angstr = 0.52917721   # in AA
+BOHR_to_angstr = 0.52917721  # in AA
```
Comment (mahrossi) [inline gensec/customPrecon.py:66]: let's make all variables non capitalized.

File: gensec/customPrecon.py [hunk 5]
```diff
@@ -48,8 +49,8 @@ def Kabsh_rmsd(atoms, initial, molindixes, removeHs=False):
     coords1 = np.dot(coords1, K)
     rmsd_kabsh = 0.0
     for v, w in zip(coords1, coords2):
-        rmsd_kabsh += sum([(v[i] - w[i])**2.0 for i in range(len(coords1[0]))])
-    return np.sqrt(rmsd_kabsh/len(coords1))
+        rmsd_kabsh += sum([(v[i] - w[i]) ** 2.0 for i in range(len(coords1[0]))])
+    return np.sqrt(rmsd_kabsh / len(coords1))
 
 
 #    VALUES for ALPHA-vdW and C6_vdW:
```
Comment (mahrossi) [inline gensec/customPrecon.py:56]: This comment should not be here. Any actual relevant information that is included in this comment should be summarized in the docstring.

File: gensec/customPrecon.py [hunk 6]
```diff
@@ -62,100 +63,335 @@ def Kabsh_rmsd(atoms, initial, molindixes, removeHs=False):
...
+    "Bk": 1985.82,
+    "Cf": 1891.92,
+    "Es": 1851.1,
+    "Fm": 1787.07,
+    "Md": 1701.0,
+    "No": 1578.18,
+}
 
 
 # Preambule from Lindh.py pthon sctipt
```
Comment (mahrossi) [inline gensec/customPrecon.py:286]: what is this comment?

File: gensec/customPrecon.py [hunk 7]
```diff
@@ -62,100 +63,335 @@ def Kabsh_rmsd(atoms, initial, molindixes, removeHs=False):
...
 HUGE = 1e10
 
-ABOHR = 0.52917721 # in AA
-HARTREE = 27.211383 # in eV
+ABOHR = 0.52917721  # in AA
```
Comment (mahrossi) [inline gensec/customPrecon.py:289]: variables capitalized again -- and some conversions had already been defined! please make it uniform.

File: gensec/customPrecon.py [hunk 8]
```diff
@@ -167,7 +403,6 @@ def vector_separation(cell_h, cell_ih, qi, qj):
     # i-PI Copyright (C) 2014-2015 i-PI developers
```
Comment (mahrossi) [inline gensec/customPrecon.py:403]: copyright can be in the docstring and this is not a file anymore.

File: gensec/customPrecon.py [hunk 9]
```diff
@@ -194,146 +429,156 @@ def vector_separation(cell_h, cell_ih, qi, qj):
...
-    dij = np.dot(cell_h, sij).T         # back to i-pi shape
+    dij = np.dot(cell_h, sij).T  # back to i-pi shape
     rij = np.linalg.norm(dij, axis=1)
 
     return dij, rij
-    
-    
+
 
 def vdwHessian(atoms):
```
Comment (mahrossi) [inline gensec/customPrecon.py:438]: docstring missing

File: gensec/customPrecon.py [hunk 10]
```diff
@@ -194,146 +429,156 @@ def vector_separation(cell_h, cell_ih, qi, qj):
...
+    # atoms = read(options.inputfile, format = options.formatfile)
 
-    N  = len(atoms)
+    N = len(atoms)
     coordinates = atoms.get_positions()
     atom_names = atoms.get_chemical_symbols()
     cell_h = atoms.get_cell()[:]
     cell_ih = atoms.get_reciprocal_cell()[:]
 
     def calculate_vdW(i, j, rij, coordinates, atom_names):
```
Comment (mahrossi) [inline gensec/customPrecon.py:448]: not sure I understand the identention here and these nested definitions. why is this like this?

Comment (mahrossi) [reply gensec/customPrecon.py:448]: make sure you take a careful look here

File: gensec/customPrecon.py [hunk 11]
```diff
@@ -194,146 +429,156 @@ def vector_separation(cell_h, cell_ih, qi, qj):
...
         hessian[3 * i + 1, :] = stack[1]
         hessian[3 * i + 2, :] = stack[2]
-        
+
     # for ind in range(len(hessian)):
-        # hessian[ind, ind] =  hessian[ind, ind] + 0.5
-    #hessian[ind, ind] = -np.sum(hessian[ind])
+    # hessian[ind, ind] =  hessian[ind, ind] + 0.5
+    # hessian[ind, ind] = -np.sum(hessian[ind])
     return hessian
```
Comment (mahrossi) [inline gensec/customPrecon.py:512]: I don't see the point of nested definitions... this needs cleaning.

File: gensec/customPrecon.py [hunk 12]
```diff
@@ -194,146 +429,156 @@ def vector_separation(cell_h, cell_ih, qi, qj):
...
 
 
 def ExpHessian(atoms, mu=1, A=1):
```
Comment (mahrossi) [inline gensec/customPrecon.py:515]: docstring missing.

File: gensec/customPrecon.py [hunk 13]
```diff
@@ -194,146 +429,156 @@ def vector_separation(cell_h, cell_ih, qi, qj):
...
     for ind in range(len(hessian)):
         hessian[ind, ind] = -np.sum(hessian[ind])
-    #dd = "/home/damaksimovda/Insync/da.maksimov.da@gmail.com/GoogleDrive/PhD/gensec/examples/Cu_EMT/generate/0000000001/"
-    #np.savetxt(os.path.join(dd, "expHessian.hes"), hessian)
+    # dd = "/home/damaksimovda/Insync/da.maksimov.da@gmail.com/GoogleDrive/PhD/gensec/examples/Cu_EMT/generate/0000000001/"
+    # np.savetxt(os.path.join(dd, "expHessian.hes"), hessian)
     return hessian
 
 
 def ExpHessian_P(atoms, mu=1, A=1):
```
Comment (mahrossi) [inline gensec/customPrecon.py:554]: docstring missing

File: gensec/customPrecon.py [hunk 14]
```diff
@@ -415,38 +662,48 @@ def exponent(self, AB, i, j):
         least the atom types within the system.
         """
         raise NotImplementedError()
+
     def Rcut(self, max_exponent):
```
Comment (mahrossi) [inline gensec/customPrecon.py:666]: is this an empty def?

File: gensec/customPrecon.py [hunk 15]
```diff
@@ -544,33 +806,39 @@ def all_within(self, Rcut, add_base_PE=False):
...
+                            abcs.add((i + ii, j + jj, k + kk))
+
+        def _norm_of_abc(abc):
+            return norm(self.latvec(abc))
 
-        def _norm_of_abc(abc): return norm(self.latvec(abc))
         return sorted(abcs, key=_norm_of_abc)
 
+
 def get_pairs(atoms1, atoms2, Rcut, use_scipy=True):
```
Comment (mahrossi) [inline gensec/customPrecon.py:837]: docstring missing

File: gensec/customPrecon.py [hunk 16]
```diff
@@ -652,19 +920,20 @@ def __init__(self, bra, atom, damper, max_sing_thres):
...
                 abc = self.abcs[a]
                 Rvec = self.lat.latvec(abc)
                 assert np.allclose(vec, Rvec + self.atom[j] - self.atom[i])
                 damp = damper.exponent(norm(vec), i, j)
                 proc_partners.append((j, abc, damp))
-            proc_partners.sort(key=operator.itemgetter(2)) # sort by damp
+            proc_partners.sort(key=operator.itemgetter(2))  # sort by damp
             self.pairs.append(proc_partners)
 
     def getvec(self, iabc):
```
Comment (mahrossi) [inline gensec/customPrecon.py:939]: docstring missing

File: gensec/customPrecon.py [hunk 17]
```diff
@@ -920,162 +1201,185 @@ def model_matrix(bra, atom, builder, damper, thres, logfile=None):
 # Obviously, this is not the most efficient way to do it.  But at least,
```
Comment (mahrossi) [inline gensec/customPrecon.py:1201]: This comment should rather be in a docstring

File: gensec/customPrecon.py [hunk 18]
```diff
@@ -920,162 +1201,185 @@ def model_matrix(bra, atom, builder, damper, thres, logfile=None):
...
 # it works...
 
+
 def _dd_matmat(val_shape, dval_di, di_dvar):
```
Comment (mahrossi) [inline gensec/customPrecon.py:1205]: docstring missing

File: gensec/customPrecon.py [hunk 19]
```diff
@@ -920,162 +1201,185 @@ def model_matrix(bra, atom, builder, damper, thres, logfile=None):
...
     assert val_shape == np.shape(dval_di)[:val_rank]
     di_shape = np.shape(dval_di)[val_rank:]
     di_rank = len(di_shape)
     assert di_shape == np.shape(di_dvar)[:di_rank]
-    axes1 = list(range(val_rank, val_rank+di_rank))
+    axes1 = list(range(val_rank, val_rank + di_rank))
     return np.tensordot(dval_di, di_dvar, (axes1, list(range(di_rank))))
 
+
 def _dd_broadcast(val, dval):
```
Comment (mahrossi) [inline gensec/customPrecon.py:1215]: docstring missing

File: gensec/customPrecon.py [hunk 20]
```diff
@@ -920,162 +1201,185 @@ def model_matrix(bra, atom, builder, damper, thres, logfile=None):
...
     val_rank = len(np.shape(val))
     assert np.shape(val) == np.shape(dval)[:val_rank]
     dval_rank = len(np.shape(dval))
-    newshape = np.shape(val) + (dval_rank-val_rank)*(1,)
+    newshape = np.shape(val) + (dval_rank - val_rank) * (1,)
     return np.reshape(val, newshape)
 
+
 def dd_sum(*arg_ds):
```
Comment (mahrossi) [inline gensec/customPrecon.py:1223]: docstring missing

File: gensec/customPrecon.py [hunk 21]
```diff
@@ -920,162 +1201,185 @@ def model_matrix(bra, atom, builder, damper, thres, logfile=None):
...
-    res = np.float(0.)
+    res = np.float(0.0)
     dres = np.zeros(shape)
     for arg, darg in arg_ds:
         res += np.asarray(arg)
         dres += np.asarray(darg)
     return res, dres
 
+
 def dd_mult(vec_d, fac):
```
Comment (mahrossi) [inline gensec/customPrecon.py:1233]: docstring missing for ALL "dd" functions. Please address all.

File: gensec/customPrecon.py (hunk 17 above)
Comment (mahrossi) [reply gensec/customPrecon.py:1201]: On a second thought: Why do you need these functions? I don't understand them. If they are really needed, a reason should be provided (please let me know what is the reason).

**Review by mahrossi — COMMENTED** [review]
I only got to line 1000 of the original precon.py . Continuing later but lots of points to be addressed.

File: gensec.py (hunk 1 above)
Comment (maksimovdmitrii) [reply gensec.py:89]: All the routines for generation and optimization rewritten and moved to the protocols.py
The main gensec.py will be much shorter.

File: gensec.py (hunk 2 above)
Comment (maksimovdmitrii) [reply gensec.py:25]: Erased

File: gensec/customPrecon.py (hunk 3 above)
Comment (maksimovdmitrii) [reply gensec/customPrecon.py:29]: Added

File: gensec/customPrecon.py (hunk 4 above)
Comment (maksimovdmitrii) [reply gensec/customPrecon.py:66]: Done

**Pull request closed without merging**